controllers/cnbi/controller.go

```diff
@@ -112,42 +110,36 @@ func (r *CustomNBImageReconciler) SetupWithManager(mgr ctrl.Manager) error {
 		Complete(r)
 }
 
-// Force object status update. Returns a reconcile result
-func (r *CustomNBImageReconciler) UpdateStatusNow(ctx context.Context, originalErr error) (ctrl.Result, error) {
+func (r *CustomNBImageReconciler) updateStatus(ctx context.Context, nn types.NamespacedName, status *meteorv1alpha1.CustomNBImageStatus) error {
 	logger := log.FromContext(ctx)
-	if err := r.Status().Update(ctx, r.CNBi); err != nil {
-		logger.WithValues("reason", err.Error()).Info("Unable to update status, retrying")
-		return ctrl.Result{Requeue: true}, nil
-	}
-	if originalErr != nil {
-		return ctrl.Result{RequeueAfter: 10 * time.Second}, originalErr
-	} else {
-		return ctrl.Result{}, nil
-	}
-}
+	logger.Info(("updating status of CustomNotebookImage"), "status", r.CNBi.Status, "newStatus", status)
 
-// Set status condition helper
-func (r *CustomNBImageReconciler) SetCondition(kind, name string, status metav1.ConditionStatus, reason, message string) {
-	meta.SetStatusCondition(&r.CNBi.Status.Conditions, metav1.Condition{
-		Type:    kind + cases.Title(language.Und).String(name),
-		Status:  status,
-		Reason:  reason,
-		Message: message,
-	})
+	if err := retry.RetryOnConflict(retry.DefaultRetry, func() error {
+		original := &meteorv1alpha1.CustomNBImage{}
+		if err := r.Get(ctx, nn, original); err != nil {
+			return err
+		}
+		original.Status = *status
+		if err := r.Client.Status().Update(ctx, original); err != nil {
+			return err
+		}
+		return nil
+	}); err != nil {
+		return fmt.Errorf("failed to update status of CNBi %s/%s: %v", nn.Namespace, nn.Name, err)
+	}
+	return nil
```

Comment:
I wonder if this might not be the root of (or at least contributing to) the failing test.
The doc for RetryOnConflict says :
```
// RetryOnConflict is used to make an update to a resource when you have to worry about
// conflicts caused by other code making unrelated updates to the resource at the same
// time. fn should fetch the resource to be modified, make appropriate changes to it, try
// to update it, and return (unmodified) the error from the update function. On a
```
But here, we get the resource at the start of the reconcile loop, mutate stuff (conditions etc) then re-get the resource and try to modify. Should we instead wrap the whole reconcile function into RetryOnConflict ?

Reply:
ja, if we return an err from Reconcile() it will be requeued automatically. If any of the owned resources changes, this controller will reconcile, see line 99

Reply:
... I am still thinking if line 92 is correct...

Reply:
> ja, if we return an err from Reconcile() it will be requeued automatically. If **any of the owned resources changes**, this controller will reconcile, see line 99

By owned resources you mean PipelineRun with `ownerRef` ? Line 99 will only requeue reconciliation if `err != nil` AFAICT.